fem/fe/fe_rbf.cpp

```diff
@@ -0,0 +1,1469 @@
+                                   const double faceFactor)
+   : KernelFiniteElement(D,
+                         TensorBasisElement::GetTensorProductGeometry(D),
+                         TensorBasisElement::Pow(numPointsD, D),
+                         intOrder*numPointsD,
+                         FunctionSpace::Qk),
+#ifndef MFEM_THREAD_SAFE
+     x_scr(D),
+     y_scr(D),
+     dy_scr(D),
+     dr_scr(D),
+     ddr_scr(D, D),
+#endif
+     numPointsD(numPointsD),
+     h(h),
+     faceFactor(faceFactor),
+     rbf(RBFType::GetRBF(rbfType)),
+     distance(DistanceMetric::GetDistance(D, distNorm))
+{
+   InitializeGeometry();
+}
+
+void RBFFiniteElement::GetCompactIndices(const Vector &ip,
+                                         int (&indices)[3][2]) const
+{
+   // This assumes the points lie on vertices of equally-spaced Cartesian grid
+   // Should work for any Lp norm, as the regions are assumed to be square
+   for (int d = 0; d < dim; ++d)
+   {
+      indices[d][0] = std::max(int(ceil((ip(d) - radPhys) / delta)), 0);
+      indices[d][1] = std::min(int(floor((ip(d) + radPhys) / delta)), numPointsD - 1);
+   }
+
+   // Fill remainder of indices with zeros so we don't have to specialize dim
+   for (int d = dim; d < 3; ++d)
+   {
+      indices[d][0] = 0;
+      indices[d][1] = 0;
+   }
+}
+
+void RBFFiniteElement::GetGlobalIndices(const Vector &ip,
+                                        int (&indices)[3][2]) const
+{
+   // Make the indices range through all the points
+   for (int d = 0; d < dim; ++d)
+   {
+      indices[d][0] = 0;
+      indices[d][1] = numPointsD - 1;
+   }
+}
+
+void RBFFiniteElement::GetTensorIndices(const Vector &ip,
+                                        int (&indices)[3][2]) const
+{
+   if (isCompact) { GetCompactIndices(ip, indices); }
+   else { GetGlobalIndices(ip, indices); }
+}
+
+void RBFFiniteElement::DistanceVec(const int i,
+                                   const Vector &x,
+                                   Vector &y) const
+{
+   switch (dim)
+   {
+      case 1:
+         y(0) = (x(0) - Nodes.IntPoint(i).x) * hPhysInv;
+         break;
+      case 2:
+         y(0) = (x(0) - Nodes.IntPoint(i).x) * hPhysInv;
+         y(1) = (x(1) - Nodes.IntPoint(i).y) * hPhysInv;
+         break;
+      case 3:
+         y(0) = (x(0) - Nodes.IntPoint(i).x) * hPhysInv;
+         y(1) = (x(1) - Nodes.IntPoint(i).y) * hPhysInv;
+         y(2) = (x(2) - Nodes.IntPoint(i).z) * hPhysInv;
+         break;
+      default:
+         MFEM_ABORT("invalid dimension: " << dim);
+   }
+}
+
+void RBFFiniteElement::InitializeGeometry()
+{
+   MFEM_ASSERT(numPointsD >= 2, "must have at least 2 points for rbf to work");
+   double deltaD = 1.0 / static_cast<double>
+                   (numPointsD); // discontinuous: points dx/2 from face
+   double factorD = deltaD * faceFactor;
+   double deltaC = 1.0 / (static_cast<double>(numPointsD) -
+                          1); // continuous: points on face
```

Comment:
Strange formatting. The comment can be above the line, and be more descriptive.

Reply:
Done.